### PR TITLE
PERF: Improve SeriesGroupBy.value_counts performances with categorical values.

### LINKED ISSUE #46202: BUG: very slow groupby(col1)[col2].value_counts() for columns of type 'category'
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
# EXAMPLE 1

import numpy as np
import pandas as pd
import time

t0 = time.time()
col1_possible_values = ["".join(np.random.choice(list("ABCDEFGHIJKLMNOPRSTUVWXYZ"), 20)) for _ in range(700000)]
col2_possible_values = ["".join(np.random.choice(list("ABCDEFGHIJKLMNOPRSTUVWXYZ"), 10)) for _ in range(860)]

col1_values = np.random.choice(col1_possible_values, size=13000000, replace=True)
col2_values = np.random.choice(col2_possible_values, size=13000000, replace=True)

sample_df = pd.DataFrame(zip(col1_values, col2_values), columns=["col1", "col2"])
print(time.time()-t0)

t0 = time.time()
processed_df = sample_df.groupby("col1")["col2"].value_counts().unstack()
print(time.time()-t0)

# EXAMPLE 2

import numpy as np
import pandas as pd
import time

t0 = time.time()
col1_possible_values = ["".join(np.random.choice(list("ABCDEFGHIJKLMNOPRSTUVWXYZ"), 20)) for _ in range(700000)]
col2_possible_values = ["".join(np.random.choice(list("ABCDEFGHIJKLMNOPRSTUVWXYZ"), 10)) for _ in range(860)]

col1_values = np.random.choice(col1_possible_values, size=13000000, replace=True)
col2_values = np.random.choice(col2_possible_values, size=13000000, replace=True)

sample_df = pd.DataFrame(zip(col1_values, col2_values), columns=["col1", "col2"])
sample_df['col2'] = sample_df['col2'].astype('category')
print(time.time()-t0)

t0 = time.time()
processed_df = sample_df.groupby("col1")["col2"].value_counts().unstack()
print(time.time()-t0)
```

### Issue Description

I have 2022 Macbook Pro M1 Pro, pandas 1.4.1, numpy 1.22.2

I noticed significant performance drop when trying to perform

sample_df.groupby("col1")["col2"].value_counts().unstack()

when col2 or both col1 and col2 are of type 'category' instead of default type 'object'. Operation in EXAMPLE 1 runs around ~25 seconds on my computer (similar for 2019 Macbook Pro with Intel processor). In EXAMPLE 2 I have run the operation for more than 20 minutes and it still did not finish (on 2019 Macbook Pro with Intel processor running time is similar for EXAMPLE 1 and EXAMPLE 2).

Moreover, for M1 in EXAMPLE 1, the peak memory usage is around 8-10GB while for EXAMPLE 2 it well exceeds 30GB. 

### Expected Behavior

on M1:

EXAMPLE 1 and EXAMPLE 2 should perform roughly the same as they do on Intel processor.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.9.10.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.3.0
Version          : Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.4.1
numpy            : 1.22.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 22.0.3
setuptools       : 60.9.3
Cython           : 0.29.28
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3
jinja2           : 3.0.3
IPython          : 8.0.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.5.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.8.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
None

</details>

### PR DESCRIPTION
**Pull request by LucasG0**
- [X] closes #46202
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

The performance issue comes from the fact the categorical path relies on `GroupBy.apply`. Actually, `SeriesGroupby.value_counts` and `DataFrameGroupBy.value_counts` have different implementations, and I feel they should both rely on a single one. Typically, the Series implementation looks more complicated than the DataFrame one (which relies on `GroupBy.size`), this is why this change creates a `DataFrameGroupBy` object to rely on the DataFrame implementation. Ideally, I think the entire Series implementation could rely on the DataFrame one (not only the categorical path). Also, I have a small doubt whether creating a new `DataFrameGroupBy` object out of a `SeriesGroupBy` one is a good practice or not, so if there is a better way to do here feel free to suggest it.

Note I did not add a test in this PR as it fixes performance only, but I manually tested and confirmed the performance improvement.

### CODE REVIEW

**Comment by jreback** [comment]
do we have asv's for this? can you show.

**Comment by LucasG0** [comment]
> do we have asv's for this? can you show. (if not, or they are not representative of the OP can you add)

It seems there is no asv for this. I will add one once I succeed to run the benchmark suite.

**Comment by LucasG0** [comment]
After benchmarking, I noticed that the current `DataFrameGroupBy.value_counts` implementation might be slower then the `SeriesGroupBy.value_counts` with a high number of categories. Here are the results I got for the two benchmarks I added (cf this [commit](https://github.com/pandas-dev/pandas/pull/46940/commits/cd20580c27892d22098e0027afa6738dca3ca338)).

benchmarks

It shows a x30 performance boost for a small number of categories, and a x2 slow down with many categories.
I also noticed a 10x boost on OP's example (with a df size of 10_000 instead of 13_000_000).

I feel this fix is still interesting considering the slow down for many categories is negligible regarding the boost introduced for a small number of categories. I guess it depends on how many categories people use in general. @jreback do you have an opinion on it ?

Also, I profiled the `DataFrameGroupBy.value_counts` implementation to see where the time is spent, and it is mainly during sorting and `_reindex_output`.

snakeviz

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -609,12 +609,35 @@ def value_counts(
 
         names = self.grouper.names + [self.obj.name]
 
-        if is_categorical_dtype(val.dtype) or (
-            bins is not None and not np.iterable(bins)
-        ):
+        if is_categorical_dtype(val.dtype):
```
Comment (jreback) [inline pandas/core/groupby/generic.py:612]: can we just call the appropriate implementation. creating a new grouper and calling like this just obfuscatest he code. would really prefer simply to have a single implementation and just call it.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (LucasG0) [reply pandas/core/groupby/generic.py:612]: I understand why it it obfuscates the code, and I wonder what would be the approriate implementation here. Do you mean introducing a method `_groupby_value_counts` containing most of the current DataFrame implementation logic, that would be called by both Series and DataFrame implementation ?

Comment (LucasG0) [reply pandas/core/groupby/generic.py:612]: @jreback could you please elaborate what you think would be the approriate implementation to call here ?

Comment (jreback) [reply pandas/core/groupby/generic.py:612]: need to refactor the value counts to a separate function then just call it for frame and series

**Comment by LucasG0** [comment]
I am no longer working on this pull request, I unassigned me of https://github.com/pandas-dev/pandas/issues/46202.
I think the follow up on this issue would be to have a common implementation for both `SeriesGroupBy` and `DataFrameGroupBy`, but this refactoring seems to be not trivial, cf [this comment](https://github.com/pandas-dev/pandas/pull/46940#discussion_r892569471).

**Comment by mroeschke** [comment]
Thanks for your efforts so far @LucasG0. I'm going to close this PR as it has gone stale, but can reopen if anyone wants to tackle in the future

**Pull request closed without merging**